### PR TITLE
Hook up Ledger Wallet Integration

### PR DESCRIPTION
**Pull request by brandonlehmann**
- Allow for using the LedgerNote library instead of CryptoNote library if LedgerTransport is provided
- Add more WalletError num entries for Ledger based operations
- Stop creating a new instance in the CryptoWrapper every time we call as we need a state mechanism
  - Caches the configuration and the instance and if the config changes regenerates the instance
- Add support for creating a new wallet using a Ledger Device
- Add support for opening a wallet file that relies on a Ledger Device (or kicks back an error)
- Add support for importWalletFromLedger
- Add CI/CD tests for testing against a Ledger device
- Add ability for Test.test() method to handle more than just true/false responses such that we can now list 'skipped' tests
- Check to verify that the Ledger device connected matches the wallet file opened

image

### CODE REVIEW

**Comment by RocksteadyTC** [comment]
yesssss

File: lib/WalletBackend.ts [hunk 1]
```diff
@@ -672,6 +700,91 @@ export class WalletBackend extends EventEmitter {
...
+     * const [wallet, err] = await WB.WalletBackend.importWalletFromLedger(daemon, 100000, {
+     *     ledgerTransport: transport
+     * });
+     *
+     * if (err) {
+     *      console.log('Failed to load wallet: ' + err.toString());
+     * }
+     * ```
+     *
+     * @param daemon        An implementation of the Daemon interface.
```
Comment (zpalmtree) [inline lib/WalletBackend.ts:724]: While this comment is copied from the others, it's no longer accurate. We got rid of the `IDaemon` interface and now they just have the one daemon class they can supply.

File: lib/WalletBackend.ts [hunk 2]
```diff
@@ -672,6 +700,91 @@ export class WalletBackend extends EventEmitter {
...
+        } catch (e) {
+            return [undefined, new WalletError(WalletErrorCode.LEDGER_COULD_NOT_GET_KEYS)];
+        }
+
+        /* Can't sync from the current scan height, not newly created */
+        const newWallet: boolean = false;
+
+        const wallet = await WalletBackend.init(
+            merged, daemon, await address.address(), scanHeight, newWallet,
+            address.view.privateKey, ''.padEnd(64, ''),
```
Comment (zpalmtree) [inline lib/WalletBackend.ts:782]: Should this not be `''.padEnd(64, '0')`?

I would prefer `'0'.repeat(64)` if that is the intended code.

File: lib/WalletBackend.ts [hunk 3]
```diff
@@ -3284,4 +3428,14 @@ export class WalletBackend extends EventEmitter {
         /* We're done. */
         this.currentlyOptimizing = false;
     }
+
+    private async isLedgerRequired(): Promise<boolean> {
+        const [privateSpendKey] = await this.getPrimaryAddressPrivateKeys();
+
+        return !this.subWallets.isViewWallet && privateSpendKey === ''.padEnd(64, '0')
```
Comment (zpalmtree) [inline lib/WalletBackend.ts:3435]:
```suggestion
        return !this.subWallets.isViewWallet && privateSpendKey === '0'.repeat(64);
```

File: lib/WalletBackend.ts [hunk 4]
```diff
@@ -788,7 +901,22 @@ export class WalletBackend extends EventEmitter {
...
+            await CryptoUtils(merged).fetchKeys();
+
+            const ledgerAddress = CryptoUtils(merged).address;
+
+            if (ledgerAddress) {
+                address = ledgerAddress;
+            } else {
+                throw new Error('Could not create wallet from Ledger transport');
+            }
+
```
Comment (zpalmtree) [inline lib/WalletBackend.ts:918]:
```suggestion
```

File: lib/WalletError.ts [hunk 5]
```diff
@@ -482,6 +494,18 @@ export enum WalletErrorCode {
 
     /* Prepared transaction cannot be found, perhaps wallet application has been restarted */
     PREPARED_TRANSACTION_NOT_FOUND = 58,
+
+    /* A Ledger transport is required */
+    LEDGER_TRANSPORT_REQUIRED = 59,
```
Comment (zpalmtree) [inline lib/WalletError.ts:499]: Could we add the missing error from https://github.com/turtlecoin/turtlecoin/blob/development/src/errors/Errors.h#L239 here to ensure the error code values match up?

File: tests/tests.ts [hunk 6]
```diff
@@ -733,13 +761,176 @@ async function roundTrip(
...
+
+            return TestStatus.FAIL;
+        },
+            'Fail to open Ledger wallet without Ledger transport',
+            'Test passed',
+            'Test failed',
+            'Ledger tests skipped');
+
+        await tester.test(async () => {
+                if (skipLedgerTests) {
```
Comment (zpalmtree) [inline tests/tests.ts:872]: Indentation is wrong on this function

File: tests/tests.ts [hunk 7]
```diff
@@ -733,13 +761,176 @@ async function roundTrip(
...
+
+            return TestStatus.FAIL;
+        },
+            'Fail to create subwallet for Ledger based wallet',
+            'Test passed',
+            'Test failed',
+            'Ledger tests skipped')
+
+        await tester.test(async () => {
+                if (skipLedgerTests) {
```
Comment (zpalmtree) [inline tests/tests.ts:914]: Indentation is wrong here

File: lib/WalletBackend.ts (hunk 1 above)
Comment (brandonlehmann) [reply lib/WalletBackend.ts:724]: @param defines method parameters, all of these have daemon as the first parameter -- are you somehow stating that they don't supply a daemon as the parameter although it's the first parameter in the method statement and is required?

File: lib/WalletBackend.ts (hunk 2 above)
Comment (brandonlehmann) [reply lib/WalletBackend.ts:782]: Yeah, that's a typo.

File: lib/WalletError.ts (hunk 5 above)
Comment (brandonlehmann) [reply lib/WalletError.ts:499]: Sure, I'll add that to the scope of this PR ;)

File: lib/WalletBackend.ts (hunk 1 above)
Comment (zpalmtree) [reply lib/WalletBackend.ts:724]: No, I'm saying Daemon isn't an interface, it's a class

**Review by zpalmtree — APPROVED** [review]

**Comment by zpalmtree** [comment]
Sorry, thought I had merged this

**Comment by brandonlehmann** [comment]
Thank you kindly sir.

**Pull request merged**